File: 04-Data-Structures/code/typescript/src/indexedHeap.ts
import { DSError } from "./errors.js";

/**
 * Min-priority-queue keyed by an opaque string id, supporting O(log n)
 * `decreaseKey` via an id -> heap-position index (the classic Dijkstra/Prim
 * building block).
 */
export class IndexedHeap {
  private heap: string[] = [];
  private pos: Map<string, number> = new Map();
  private priority: Map<string, number> = new Map();

  size(): number {
    return this.heap.length;
  }

  contains(id: string): boolean {
    return this.pos.has(id);
  }

  push(id: string, priority: number): void {
    if (this.pos.has(id)) throw new DSError("invalid", `id ${id} already present`);
    this.heap.push(id);
    const i = this.heap.length - 1;
    this.pos.set(id, i);
    this.priority.set(id, priority);
    this.bubbleUp(i);
  }

  decreaseKey(id: string, newPriority: number): void {
    if (!this.pos.has(id)) throw new DSError("missing", `id ${id} not found`);
    const current = this.priority.get(id) as number;
    if (newPriority > current) throw new DSError("invalid", "newPriority must be <= current priority");
    this.priority.set(id, newPriority);
    this.bubbleUp(this.pos.get(id) as number);
  }

  pop(): string {
    if (this.heap.length === 0) throw new DSError("empty", "pop from empty heap");
    const top = this.heap[0];
    const last = this.heap.pop() as string;
    this.pos.delete(top);
    this.priority.delete(top);
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this.pos.set(last, 0);
      this.bubbleDown(0);
    }
    return top;
  }

  private less(i: number, j: number): boolean {
    return (this.priority.get(this.heap[i]) as number) < (this.priority.get(this.heap[j]) as number);
  }

  private swap(i: number, j: number): void {
    const a = this.heap[i];
    const b = this.heap[j];
    this.heap[i] = b;
    this.heap[j] = a;
    this.pos.set(b, i);
    this.pos.set(a, j);
  }

  private bubbleUp(i: number): void {
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.less(i, parent)) {
        this.swap(i, parent);
        i = parent;
      } else break;
    }
  }

  private bubbleDown(i: number): void {
    const n = this.heap.length;
    while (true) {
      let smallest = i;
      const l = 2 * i + 1;
      const r = 2 * i + 2;
      if (l < n && this.less(l, smallest)) smallest = l;
      if (r < n && this.less(r, smallest)) smallest = r;
      if (smallest === i) break;
      this.swap(i, smallest);
      i = smallest;
    }
  }

  checkInvariants(): void {
    for (let i = 1; i < this.heap.length; i++) {
      const parent = (i - 1) >> 1;
      if (this.less(i, parent)) throw new DSError("invalid", "heap property violated");
    }
    if (this.pos.size !== this.heap.length) throw new DSError("invalid", "position map size mismatch");
    for (let i = 0; i < this.heap.length; i++) {
      if (this.pos.get(this.heap[i]) !== i) throw new DSError("invalid", "position map entry mismatch");
    }
  }
}

File: 04-Data-Structures/code/python/seb_ds/indexed_heap.py
"""Min-priority-queue keyed by an opaque string id. Mirrors `typescript/src/indexedHeap.ts`."""

from __future__ import annotations

from typing import Dict, List

from .errors import DSError


class IndexedHeap:
    def __init__(self) -> None:
        self._heap: List[str] = []
        self._pos: Dict[str, int] = {}
        self._priority: Dict[str, float] = {}

    def size(self) -> int:
        return len(self._heap)

    def contains(self, id_: str) -> bool:
        return id_ in self._pos

    def push(self, id_: str, priority: float) -> None:
        if id_ in self._pos:
            raise DSError("invalid", f"id {id_} already present")
        self._heap.append(id_)
        i = len(self._heap) - 1
        self._pos[id_] = i
        self._priority[id_] = priority
        self._bubble_up(i)

    def decrease_key(self, id_: str, new_priority: float) -> None:
        if id_ not in self._pos:
            raise DSError("missing", f"id {id_} not found")
        current = self._priority[id_]
        if new_priority > current:
            raise DSError("invalid", "newPriority must be <= current priority")
        self._priority[id_] = new_priority
        self._bubble_up(self._pos[id_])

    def pop(self) -> str:
        if not self._heap:
            raise DSError("empty", "pop from empty heap")
        top = self._heap[0]
        last = self._heap.pop()
        del self._pos[top]
        del self._priority[top]
        if self._heap:
            self._heap[0] = last
            self._pos[last] = 0
            self._bubble_down(0)
        return top

    def _less(self, i: int, j: int) -> bool:
        return self._priority[self._heap[i]] < self._priority[self._heap[j]]

    def _swap(self, i: int, j: int) -> None:
        a, b = self._heap[i], self._heap[j]
        self._heap[i], self._heap[j] = b, a
        self._pos[b] = i
        self._pos[a] = j

    def _bubble_up(self, i: int) -> None:
        while i > 0:
            parent = (i - 1) >> 1
            if self._less(i, parent):
                self._swap(i, parent)
                i = parent
            else:
                break

    def _bubble_down(self, i: int) -> None:
        n = len(self._heap)
        while True:
            smallest = i
            l = 2 * i + 1
            r = 2 * i + 2
            if l < n and self._less(l, smallest):
                smallest = l
            if r < n and self._less(r, smallest):
                smallest = r
            if smallest == i:
                break
            self._swap(i, smallest)
            i = smallest

    def check_invariants(self) -> None:
        for i in range(1, len(self._heap)):
            parent = (i - 1) >> 1
            if self._less(i, parent):
                raise DSError("invalid", "heap property violated")
        if len(self._pos) != len(self._heap):
            raise DSError("invalid", "position map size mismatch")
        for i, id_ in enumerate(self._heap):
            if self._pos[id_] != i:
                raise DSError("invalid", "position map entry mismatch")
